File: [24.01]Spring-Batch/Java/src/main/java/com/example/java/itemReader/cursor/JpaPagingItemReaderJobConfiguration.java
package com.example.java.itemReader.cursor;


import jakarta.persistence.EntityManagerFactory;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.batch.core.Job;
import org.springframework.batch.core.Step;
import org.springframework.batch.core.job.builder.JobBuilder;
import org.springframework.batch.core.repository.JobRepository;
import org.springframework.batch.core.step.builder.StepBuilder;
import org.springframework.batch.item.ItemWriter;
import org.springframework.batch.item.database.JpaPagingItemReader;
import org.springframework.batch.item.database.builder.JpaPagingItemReaderBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.transaction.PlatformTransactionManager;

@Slf4j
@Configuration
@RequiredArgsConstructor
public class JpaPagingItemReaderJobConfiguration {
    private ItemWriter<Pay> jpaPagingItemWriter() {
        return list -> {
            for (Pay pay :
                    list) {
                log.info("Current Pay={}", pay);
            }
        };
    }

    @Bean
    public JpaPagingItemReader<Pay> jpaPagingItemReader(EntityManagerFactory entityManagerFactory) {
        return new JpaPagingItemReaderBuilder<Pay>()
                .name("jpaPagingItemReader")
                .pageSize(10)
                .queryString("SELECT p FROM Pay p")
                .entityManagerFactory(entityManagerFactory)
                .build();
    }

    @Bean
    public Step jpaPagingItemReaderStep(JobRepository jobRepository, PlatformTransactionManager platformTransactionManager, JpaPagingItemReader<Pay> jpaPagingItemReader) {
        log.info(">>>> jpaPagingItemReaderStep");
        return new StepBuilder("jpaPagingItemReaderStep", jobRepository)
                .<Pay, Pay>chunk(10, platformTransactionManager)
                .reader(jpaPagingItemReader)
//                .processor(processor())
                .writer(jpaPagingItemWriter())
                .build();
    }

    @Bean
    public Job jpaPagingItemReaderJob(JobRepository jobRepository, Step jpaPagingItemReaderStep) {
        log.info(">>>> jpaPagingItemReaderJob");
        return new JobBuilder("jpaPagingItemReaderJob", jobRepository)
                .start(jpaPagingItemReaderStep)
                .build();
    }
}

File: [24.01]Spring-Batch/kotlin/src/main/kotlin/com/example/kotlin/itemReader/cursor/JpaPagingItemReaderJobConfiguration.kt
package com.example.kotlin.itemReader.cursor

import jakarta.persistence.EntityManagerFactory
import org.slf4j.LoggerFactory
import org.springframework.batch.core.Job
import org.springframework.batch.core.Step
import org.springframework.batch.core.job.builder.JobBuilder
import org.springframework.batch.core.repository.JobRepository
import org.springframework.batch.core.step.builder.StepBuilder
import org.springframework.batch.item.ItemWriter
import org.springframework.batch.item.database.JpaPagingItemReader
import org.springframework.batch.item.database.builder.JpaPagingItemReaderBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.transaction.PlatformTransactionManager

@Configuration
class JpaPagingItemReaderJobConfiguration {
    private val log = LoggerFactory.getLogger(this.javaClass)!!

    @Bean
    fun jpaPagingItemReader(entityManagerFactory: EntityManagerFactory): JpaPagingItemReader<Pay> =
        JpaPagingItemReaderBuilder<Pay>()
            .name("jpaPagingItemReader")
            .pageSize(10)
            .queryString("SELECT p FROM Pay p")
            .entityManagerFactory(entityManagerFactory)
            .build()

    @Bean
    fun jpaPagingItemWriter(): ItemWriter<Pay> =
        ItemWriter { items -> items.forEach { log.info("Current Pay={}", it.toString()) } }

    @Bean
    fun jpaPagingItemReaderStep(
        jobRepository: JobRepository,
        platformTransactionManager: PlatformTransactionManager,
        jpaPagingItemReader: JpaPagingItemReader<Pay>
    ): Step {
        log.info(">>>> jpaPagingItemReaderStep")
        return StepBuilder("jpaPagingItemReaderStep", jobRepository)
            .chunk<Pay, Pay>(10, platformTransactionManager)
            .reader(jpaPagingItemReader)
            .writer(jpaPagingItemWriter())
            .build()
    }

    @Bean
    fun jpaPagingItemReaderJob(jobRepository: JobRepository, jpaPagingItemReaderStep: Step): Job {
        log.info(">>>> jpaPagingItemReaderJob")
        return JobBuilder("jpaPagingItemReaderJob", jobRepository)
            .start(jpaPagingItemReaderStep)
            .build()
    }
}